README.md
# Release

Release is a JavaScript plugin that enables the execution of an action after the user keeps the mouse pressed for a custom amount of time.

[![Build
Status](https://travis-ci.org/daviferreira/release.png)](https://travis-ci.org/daviferreira/release)

## Demos

* __[Menu](http://www.daviferreira.com/release/demo/menu.html)__
* __[Press to delete](http://www.daviferreira.com/release/demo/delete.html)__
* __[Shopping Cart](http://www.daviferreira.com/release/demo/shopping-cart.html)__

## Basic usage

First you need to attach release stylesheet and javascript files:

```html
<link rel="stylesheet" href="css/release.css">
<script src="js/release.js"></script>
```

Now you can instantiate a Release object by using the following code:

```html
<script>var release = new Release('.release', {onRelease: myFunction()});</script>
```

With the above code, all elements with the _release_ class will receive a _mouseup_ event that will execute _myFunction()__ after 3 seconds (default value).

## Initialization options

```javascript
defaults: {
    animationTime: 200,
    increaseRate: 15,
    onRelease: '',
    releaseTime: 3000
}
```

* __animationTime__: defines the interval to animate the indicator. It must be the same as the css transition timing value. Default: 200
* __increaseRate__: pixel value that will be added to the indicator dimensions when the user is pressing the mouse. Default: 15
* __onRelease__: callback to be executed when _releaseTime_ is reached. Default: none (it will just reset the indicator)
* __releaseTime__: time limit until the indicator is released. Default: 3000

## Development

Release development tasks are managed by Grunt. To install all the necessary packages, just invoke:

```bash
npm install
```

These are the available grunt tasks:

* __js__: runs jslint and jasmine tests and creates minified and concatenated versions of the script;
* __css__: runs compass and csslint
* __test__: runs jasmine tests, jslint and csslint
* __watch__: watch for modifications on script/scss files

The source files are located inside the __src__ directory. Release stylesheet was created using sass/compass, so make sure you have the __compass gem__ installed on your system.

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Test your changes to the best of your ability.
4. Update the documentation to reflect your changes if they add or changes current functionality.
5. Commit your changes (`git commit -am 'Added some feature'`)
6. Push to the branch (`git push origin my-new-feature`)
7. Create new Pull Request

## License

"THE BEER-WARE LICENSE" (Revision 42):

As long as you retain this notice you can do whatever you want with this stuff. If we meet some day, and you think this stuff is worth it, you can buy me a beer in return.


CHANGES.md
0.1.0 / 2014-01-23
==================

* Initial release


spec/helpers/util.js
function fireEvent(element, event, keyCode, ctrlKey) {
    'use strict';
    var evt = document.createEvent("HTMLEvents");
    evt.initEvent(event, true, true);
    if (keyCode) {
        evt.keyCode = keyCode;
    }
    if (ctrlKey) {
        evt.ctrlKey = true;
    }
    return !element.dispatchEvent(evt);
}


spec/release.spec.js
/*global describe, it, expect, Release, console, beforeEach, afterEach,
         jasmine, fireEvent, ReleaseIndicator, spyOn*/

describe('Indicator TestCase', function () {
    'use strict';

    beforeEach(function () {
        this.release = new Release('.release');
    });

    afterEach(function () {
        this.release = undefined;
    });

    describe('Init', function () {
        it('should have a default set of options', function () {
            expect(this.release.options).toEqual({
                animationTime: 200,
                increaseRate: 15,
                onRelease: '',
                releaseTime: 3000
            });
        });

        it('should accept custom options', function () {
            var customOptions = {
                increaseRate: 30,
                onRelease: function () {
                    console.log('release');
                }
            };
            this.release = new Release('.release', customOptions);
            expect(this.release.options).toEqual(customOptions);
        });

        it('should accept an array of HTML nodes', function () {
            var div = document.body.appendChild(document.createElement('div'));
            this.release = new Release([div]);
            expect(this.release.elements[0]).toBe(div);
            document.body.removeChild(div);
        });

        it('should accept a selector', function () {
            var div = document.createElement('div');
            div.className = 'release';
            document.body.appendChild(div);
            this.release = new Release('.release');
            expect(this.release.elements[0]).toBe(div);
            document.body.removeChild(div);
        });
    });

    describe('Event binding', function () {
        beforeEach(function () {
            jasmine.Clock.useMock();
            var div = document.createElement('div');
            div.className = 'release';
            document.body.appendChild(div);
            this.release = new Release('.release');
        });

        afterEach(function () {
            clearInterval(this.release.timer);
            this.release = undefined;
        });

        describe('mouseup', function () {
            it('should create the indicator on mousedown when it doest not exists', function () {
                spyOn(ReleaseIndicator.prototype, 'init');
                fireEvent(this.release.elements[0], 'mousedown');
                jasmine.Clock.tick(this.release.options.animationTime + 1);
                expect(ReleaseIndicator.prototype.init).toHaveBeenCalledWith(this.release.options.increaseRate,
                                                                             this.release.options.animationTime);
            });

            it('shoud animate the indicator on mousedown', function () {
                spyOn(ReleaseIndicator.prototype, 'animate');
                fireEvent(this.release.elements[0], 'mousedown');
                jasmine.Clock.tick((this.release.options.animationTime * 2) + 1);
                expect(this.release.indicator.animate).toHaveBeenCalled();
            });

            it('should lock the indicator when releaseTime is reached', function () {
                spyOn(Release.prototype, 'lock');
                fireEvent(this.release.elements[0], 'mousedown');
                jasmine.Clock.tick(this.release.options.releaseTime + 1);
                expect(this.release.lock).toHaveBeenCalled();
            });

            it('should call the default release method when a callback is not defined', function () {
                spyOn(Release.prototype, 'release');
                fireEvent(this.release.elements[0], 'mousedown');
                jasmine.Clock.tick(this.release.options.releaseTime + 1);
                expect(this.release.release).toHaveBeenCalled();
            });

            it('should execute the onRelease callback when releaseTime is reached', function () {
                var callback = jasmine.createSpy();
                this.release = new Release('.release', {
                    onRelease: callback
                });
                fireEvent(this.release.elements[0], 'mousedown');
                jasmine.Clock.tick(this.release.options.releaseTime + 1);
                expect(callback).toHaveBeenCalled();
            });

            it('should do nothing when indicator is locked', function () {
                this.release.isLocked = true;
                spyOn(ReleaseIndicator.prototype, 'init');
                fireEvent(this.release.elements[0], 'mousedown');
                jasmine.Clock.tick(this.release.options.animationTime + 1);
                expect(this.release.indicator).toBe(undefined);
            });
        });

        describe('mousedown', function () {
            it('should do nothing when indicator is locked', function () {
                fireEvent(this.release.elements[0], 'mousedown');
                jasmine.Clock.tick(this.release.options.animationTime + 1);
                clearInterval(this.release.timer);
                this.release.isLocked = true;
                spyOn(this.release.indicator, 'reset');
                fireEvent(this.release.elements[0], 'mouseup');
                expect(this.release.indicator.reset).not.toHaveBeenCalled();
            });

            it('should reset the indicator', function () {
                fireEvent(this.release.elements[0], 'mousedown');
                jasmine.Clock.tick(this.release.options.animationTime + 1);
                clearInterval(this.release.timer);
                spyOn(this.release.indicator, 'reset');
                fireEvent(this.release.elements[0], 'mouseup');
                expect(this.release.indicator.reset).toHaveBeenCalled();
            });
        });
    });
});


spec/indicator.spec.js
/*global describe, it, expect, ReleaseIndicator, console, beforeEach, afterEach,
         jasmine*/

describe('Indicator TestCase', function () {
    'use strict';

    beforeEach(function () {
        this.indicator = new ReleaseIndicator(15, 3000);
    });

    afterEach(function () {
        document.body.removeChild(document.body.querySelector('div.release-indicator'));
        this.indicator = undefined;
    });

    describe('Init', function () {
        it('should set the indicator increaseRate', function () {
            expect(this.indicator.increaseRate).toBe(15);
        });

        it('should set the indicator animationTime', function () {
            expect(this.indicator.animationTime).toBe(3000);
        });

        it('should create the indicator element', function () {
            expect(document.body.querySelector('div.release-indicator')).toBeTruthy();
        });
    });

    describe('Animate', function () {
        it('should position the indicator according to cursor position', function () {
            this.indicator.animate({pageY: 60, pageX: 20});
            expect(this.indicator.element.style.top).toBe('60px');
            expect(this.indicator.element.style.left).toBe('20px');
        });

        it('should add the release-indicator-active class when indicator is not active', function () {
            jasmine.Clock.useMock();
            this.indicator.animate({pageY: 10, pageX: 40});
            jasmine.Clock.tick(this.indicator.animationTime + 1);
            expect(this.indicator.element.className).toContain('release-indicator-active');
        });

        it('should set element dimensions when indicator is active', function () {
            this.indicator.isActive = true;
            this.indicator.animate({pageY: 10, pageX: 40});
            expect(this.indicator.element.style.width).toBe(this.indicator.increaseRate + 'px');
            expect(this.indicator.element.style.height).toBe(this.indicator.increaseRate + 'px');
        });
    });

    describe('Reset', function () {
        it('should clear element position and dimension', function () {
            this.indicator.isActive = true;
            this.indicator.animate({pageY: 10, pageX: 40});
            this.indicator.reset();
            expect(this.indicator.element.style.width).toBe('0px');
            expect(this.indicator.element.style.height).toBe('0px');
            expect(this.indicator.element.style.marginTop).toBe('');
            expect(this.indicator.element.style.marginLeft).toBe('');
            expect(this.indicator.element.style.top).toBe('');
            expect(this.indicator.element.style.left).toBe('');
        });

        it('should remove the release-indicator-active class', function () {
            this.indicator.element.classList.add('release-indicator-active');
            this.indicator.reset();
            expect(this.indicator.element.className).not.toContain('release-indicator-active');
        });

        it('should set the indicator as inactive', function () {
            this.indicator.isActive = true;
            this.indicator.reset();
            expect(this.indicator.isActive).toBe(false);
        });
    });
});


Gruntfile.js
/*global module, require*/

module.exports = function(grunt) {
    'use strict';

    var gruntConfig = {
        pkg: grunt.file.readJSON('package.json')
    };

    gruntConfig.jslint = {
        files: ['src/js/**/*.js', 'spec/**/*.js', 'Gruntfile.js'],
        directives: {
            browser: true,
            unparam: true,
            todo: true,
            debug: true
        }
    };

    gruntConfig.jasmine = {
        suite: {
            src: 'src/js/**/*.js',
            options: {
                specs: 'spec/*.spec.js',
                helpers: 'spec/helpers/*.js',
                styles: 'dist/css/*.css',
                junit: {
                    path: "reports/jasmine/",
                    consolidate: true
                },
                keepRunner: true,
                template: require('grunt-template-jasmine-istanbul'),
                templateOptions: {
                    coverage: 'reports/jasmine/coverage.json',
                    report: 'coverage'
                }
            }
        }
    };

    gruntConfig.uglify = {
        options: {
            report: 'gzip'
        },
        build: {
            src: 'src/js/release.js',
            dest: 'dist/js/<%= pkg.name %>.min.js'
        }
    };

    gruntConfig.csslint = {
        strict: {
            options: {
                'box-sizing': false,
                'import': 2
            },
            src: 'dist/css/**/*.css'
        }
    };

    gruntConfig.compass = {
        dist: {
            options: {
                sassDir: 'src/sass',
                cssDir: 'dist/css',
                outputStyle: 'compressed',
                noLineComments: true
            }
        }
    };

    gruntConfig.watch = {
        scripts: {
            files: ['src/js/**/*.js', 'spec/**/*.js', 'Gruntfile.js'],
            tasks: ['js'],
            options: {
                debounceDelay: 250
            }
        },
        styles: {
            files: 'src/sass/**/*.scss',
            tasks: ['css'],
            options: {
                debounceDelay: 250
            }
        }
    };

    gruntConfig.concat = {
        options: {
            stripBanners: true
        },
        dist: {
            src: ['src/js/indicator.js', 'src/js/release.js'],
            dest: 'dist/js/<%= pkg.name %>.js'
        }
    };

    gruntConfig.plato = {
        feed: {
            files: {
                'reports/plato': ['src/js/**/*.js']
            }
        }
    };

    grunt.initConfig(gruntConfig);

    grunt.loadNpmTasks('grunt-contrib-uglify');
    grunt.loadNpmTasks('grunt-jslint');
    grunt.loadNpmTasks('grunt-contrib-jasmine');
    grunt.loadNpmTasks('grunt-contrib-compass');
    grunt.loadNpmTasks('grunt-contrib-csslint');
    grunt.loadNpmTasks('grunt-contrib-watch');
    grunt.loadNpmTasks('grunt-contrib-concat');
    grunt.loadNpmTasks('grunt-plato');

    grunt.registerTask('test', ['jslint', 'jasmine', 'csslint']);
    grunt.registerTask('js', ['jslint', 'jasmine', 'uglify', 'concat']);
    grunt.registerTask('css', ['compass', 'csslint']);
    grunt.registerTask('default', ['js', 'css']);

};


src/js/indicator.js
/*global console, module*/

function ReleaseIndicator(increaseRate, animationTime) {
    'use strict';
    return this.init(increaseRate, animationTime);
}

(function (window, document) {
    'use strict';

    ReleaseIndicator.prototype = {

        init: function init(increaseRate, animationTime) {
            this.increaseRate = increaseRate;
            this.animationTime = animationTime;
            this.element = document.createElement('div');
            this.element.className = 'release-indicator';
            this.isActive = false;
            document.body.appendChild(this.element);
        },

        animate: function animate(e) {
            var height,
                width,
                self;
            height = (this.element.offsetHeight + this.increaseRate);
            width = (this.element.offsetWidth + this.increaseRate);
            this.element.style.top = e.pageY + 'px';
            this.element.style.left = e.pageX + 'px';
            if (!this.isActive) {
                self = this;
                this.isActive = true;
                setTimeout(function () {
                    self.element.classList.add('release-indicator-active');
                }, this.animationTime);
            } else {
                this.element.style.width = width + 'px';
                this.element.style.height = height + 'px';
                this.element.style.marginTop = (-(height / 2)) + 'px';
                this.element.style.marginLeft = (-(width / 2)) + 'px';
            }
        },

        reset: function reset() {
            this.isActive = false;
            if (this.element === undefined) {
                return;
            }
            this.element.style.width = '0';
            this.element.style.height = '0';
            this.element.style.marginTop = '';
            this.element.style.marginLeft = '';
            this.element.style.top = '';
            this.element.style.left = '';
            this.element.classList.remove('release-indicator-active');
        }

    };

}(window, document));


src/js/release.js
/*global console, module, ReleaseIndicator*/

function Release(elements, options) {
    'use strict';
    return this.init(elements, options);
}

(function (window, document) {
    'use strict';

    function extend(b, a) {
        var prop;
        if (b === undefined) {
            return a;
        }
        for (prop in a) {
            if (a.hasOwnProperty(prop) && b.hasOwnProperty(prop) === false) {
                b[prop] = a[prop];
            }
        }
        return b;
    }

    Release.prototype = {

        defaults: {
            animationTime: 200,
            increaseRate: 15,
            onRelease: '',
            releaseTime: 3000
        },

        init: function init(elements, options) {
            this.options = extend(options, this.defaults);
            this.isLocked = false;
            this.seconds = 0;
            if (typeof elements === 'string') {
                this.elements = document.querySelectorAll(elements);
            } else {
                this.elements = elements;
            }
            this.initElements();
        },

        initElements: function initElements() {
            var i;
            for (i = 0; i < this.elements.length; i += 1) {
                this.bindRelease(i);
            }
        },

        // TODO: break method
        bindRelease: function bindRelease(index) {
            var self = this,
                resetTimer = function resetTimer() {
                    clearInterval(self.timer);
                    self.seconds = 0;
                };

            this.elements[index].addEventListener('mousedown', function(e) {
                if (self.isLocked) {
                    return;
                }
                self.timer = setInterval(function () {
                    self.seconds += self.options.animationTime;
                    if (self.seconds >= self.options.releaseTime) {
                        resetTimer();
                        self.lock();
                        if (typeof self.options.onRelease === 'function') {
                            self.options.onRelease(self.elements[index]);
                        } else {
                            self.release();
                        }
                    } else {
                        if (self.indicator === undefined) {
                            self.indicator = new ReleaseIndicator(self.options.increaseRate,
                                                                  self.options.animationTime);
                        } else {
                            self.indicator.animate(e);
                        }
                    }
                }, self.options.animationTime);
            });

            document.body.addEventListener('mouseup', function() {
                if (self.isLocked) {
                    return;
                }
                resetTimer();
                if (self.indicator) {
                    self.indicator.reset();
                }
            });
        },

        release: function release() {
            this.unlock();
            this.seconds = 0;
            this.indicator.reset();
        },

        lock: function lock() {
            this.isLocked = true;
        },

        unlock: function unlock() {
            this.isLocked = false;
        }
    };

}(window, document));
